File: 380-insert-delete-getrandom-o1/RandomizedSet.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import random

class RandomizedSet(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.eleList = []
        self.eleDict = {}

    def insert(self, val):
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        if val in self.eleDict:
            return False
        self.eleList.append(val)
        self.eleDict[val] = len(self.eleList) - 1
        return True

    def remove(self, val):
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        :type val: int
        :rtype: bool
        """
        if val not in self.eleDict:
            return False

        index = self.eleDict[val]
        last = self.eleList[-1]

        self.eleList[index] = last
        self.eleDict[last] = index

        del self.eleDict[val]
        self.eleList.pop()

        return True

    def getRandom(self):
        """
        Get a random element from the set.
        :rtype: int
        """
        index = random.randint(0, len(self.eleList) - 1)
        return self.eleList[index]

obj = RandomizedSet()
obj.insert(3)
obj.remove(3)
obj.remove(0)
obj.insert(3)
obj.getRandom()
obj.remove(0)

File: 380-insert-delete-getrandom-o1/RandomizedSet.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

/*
不会，参考网上解法
set 没法常数时间随机取一个元素？
数组删除元素不是常数时间的！
*/
using namespace std;
class RandomizedSet {
private:
    vector<int> eleVector;
    unordered_map<int, int> eleMap;
public:
    /** Initialize your data structure here. */
    RandomizedSet() {

    }

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (eleMap.find(val) != eleMap.end()) return false;
        eleVector.push_back(val);
        eleMap[val] = eleVector.size() - 1;
        return true;
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (eleMap.find(val) == eleMap.end()) return false;

        eleVector[eleMap[val]] = eleVector.back();
        eleMap[eleVector.back()] = eleMap[val];

        eleVector.pop_back();
        eleMap.erase(val);
        return true;
    }

    /** Get a random element from the set. */
    int getRandom() {
        return eleVector[rand() % eleVector.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * bool param_1 = obj.insert(val);
 * bool param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
int main() {
    RandomizedSet obj = RandomizedSet();
    bool param_1 = obj.insert(1);
    bool param_2 = obj.remove(2);
    int param_3 = obj.getRandom();
    return 0;
}
